netmaster/resources/epgresource.go

```diff
@@ -0,0 +1,298 @@
+			r.Clear()
+		}
+	}()
+
+	oper := &AutoEPGOperResource{FreeEPGs: r.EPGs}
+	oper.StateDriver = r.StateDriver
+	oper.ID = r.ID
+	err = oper.Write()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Deinit the resource.
+func (r *AutoEPGCfgResource) Deinit() {
+	oper := &AutoEPGOperResource{}
+	oper.StateDriver = r.StateDriver
+	err := oper.Read(r.ID)
+	if err != nil {
+		// continue cleanup
+	} else {
+		err = oper.Clear()
+		if err != nil {
+			// continue cleanup
+		}
+	}
+
+	r.Clear()
+}
+
+// Reinit the Resource. Requires a *bitset.BitSet.
+func (r *AutoEPGCfgResource) Reinit(rsrcCfg interface{}) error {
+	var ok bool
+	prevEPGs := r.EPGs
+	r.EPGs, ok = rsrcCfg.(*bitset.BitSet)
+	if !ok {
+		return core.Errorf("Invalid type for EPG resource config")
+	}
+	err := r.Write()
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err != nil {
+			r.Clear()
+		}
+	}()
+
+	oper := &AutoEPGOperResource{}
+	oper.StateDriver = r.StateDriver
+	oper.ID = r.ID
+	err = oper.Read(r.ID)
+	if err != nil {
+		return err
+	}
+
+	prevEPGs.InPlaceSymmetricDifference(oper.FreeEPGs)
+	oper.FreeEPGs = r.EPGs
+	for i, e := prevEPGs.NextSet(0); e; i, e = prevEPGs.NextSet(i + 1) {
+		oper.FreeEPGs.Clear(i)
+	}
+
+	err = oper.Write()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Description is a description of this resource. returns AutoEPGResource.
+func (r *AutoEPGCfgResource) Description() string {
+	return AutoEPGResource
+}
+
+// GetList returns number of epgs and stringified list of epgs in use.
+func (r *AutoEPGCfgResource) GetList() (uint, string) {
+	cfg := &AutoEPGCfgResource{}
+	cfg.StateDriver = r.StateDriver
+	if err := cfg.Read(r.ID); err != nil {
+		log.Errorf("Error reading resource %s: %s", r.ID, err)
+		return 0, ""
+	}
+
+	oper := &AutoEPGOperResource{}
+	oper.StateDriver = r.StateDriver
+	if err := oper.Read(r.ID); err != nil {
+		log.Errorf("error fetching the epg resource: id %s", r.ID)
+		return 0, ""
+	}
+	oper.FreeEPGs.InPlaceSymmetricDifference(cfg.EPGs)
+
+	numEPGs := uint(0)
+	idx := uint(0)
+	startIdx := idx
+	list := []string{}
+	inRange := false
+
+	for {
+		foundValue, found := oper.FreeEPGs.NextSet(idx)
+		if !found {
+			break
+		}
+		numEPGs++
+
+		if !inRange { // begin of range
+			startIdx = foundValue
+			inRange = true
+		} else if foundValue > idx { // end of range
+			thisRange := rangePrint(startIdx, idx-1)
+			list = append(list, thisRange)
+			startIdx = foundValue
+		}
+		idx = foundValue + 1
+	}
+
+	// list end with allocated value
+	if inRange {
+		thisRange := rangePrint(startIdx, idx-1)
+		list = append(list, thisRange)
+	}
+
+	return numEPGs, strings.Join(list, ", ")
+}
+
+// Allocate a resource.
+func (r *AutoEPGCfgResource) Allocate(reqVal interface{}) (interface{}, error) {
+	oper := &AutoEPGOperResource{}
+	oper.StateDriver = r.StateDriver
+	err := oper.Read(r.ID)
+	if err != nil {
+		return nil, err
+	}
+
+	var epgID uint
+	if (reqVal != nil) && (reqVal.(uint) != 0) {
```

Comment:
since this resource allocation is only for id. we do not expect to have a specific id requested. so this may not be needed. We could straight away allocated the next free id in the bitset.